src/migrations/branch.rs

```diff
@@ -0,0 +1,216 @@
+use std::iter::FromIterator;
+use fs_err as fs;
+
+use std::ops::Index;
+use anyhow::anyhow;
+use crossterm::style::Stylize;
+use indexmap::IndexMap;
+use crate::connect::Connection;
+use crate::migrations::{Context, create, db_migration, migrate, migration};
+use crate::migrations::create::{MigrationKey, MigrationToText};
+use crate::migrations::db_migration::{DBMigration, read_all};
+use crate::migrations::dev_mode::rebase_to_schema;
+use crate::migrations::extract::DatabaseMigration;
+use crate::migrations::grammar::parse_migration;
+use crate::migrations::migrate::apply_migrations_inner;
+use crate::migrations::migration::MigrationFile;
+use crate::print;
+
+struct RebaseMigration<'a> {
+    key: MigrationKey,
+    migration: &'a DBMigration,
+    parent: Option<&'a str>,
+}
+
+impl MigrationToText for RebaseMigration<'_> {
+    type StatementsIter<'a> = std::iter::Once<&'a String> where Self: 'a;
+
+    fn key(&self) -> &MigrationKey {
+        &self.key
+    }
+
+    fn parent(&self) -> anyhow::Result<&str> {
+        if let Some(parent) = self.parent {
+            return Ok(parent)
+        }
+
+        if self.migration.parent_names.is_empty() {
+            return Ok("initial");
+        }
+
+        Ok(&self.migration.parent_names[0])
+    }
+
+    fn id(&self) -> anyhow::Result<&str> {
+        Ok(&self.migration.name)
+    }
+
+    fn statements<'a>(&'a self) -> Self::StatementsIter<'a> {
+        std::iter::once(&self.migration.script)
+    }
+}
+
+pub struct RebaseMigrations {
+    source_migrations: IndexMap<String, DBMigration>,
+    target_migrations: IndexMap<String, DBMigration>,
+}
```

Comment:
Do I understand this correctly:

`source_migrations` contains all migrations that are on the source branch,
`target_migrations` contains migrations that are on the target branch, but not on the source branch

---

If this is so, I would suggest to refactor this into:
```
pub struct RebaseMigrations {
    /// Migrations initial..base
    base_migrations: IndexMap<String, DBMigration>,
    /// Migrations base..source
    source_migrations: IndexMap<String, DBMigration>,
    /// Migrations base..target
    target_migrations: IndexMap<String, DBMigration>,
}
```
This structure would be useful in `.flatten()`.